### PR TITLE
Allow users to create forms

### PR DESCRIPTION
**Pull request by tfidfwastaken**
Closes [sc-162].

Allow users to create a form from the dashboard. Created forms are displayed on the dashboard.  
Currently, questions cannot be added, and only a non-functional view is displayed.

### CODE REVIEW

File: test/bugle_forms/handlers/form_test.clj [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+   [bugle-forms.handlers.form :as sut]
+   [bugle-forms.model.form :as form]
+   [bugle-forms.model.user :as user]
+   [clojure.test :refer [deftest is testing use-fixtures]]
+   [next.jdbc.sql :as sql]))
+
+(use-fixtures :once fixtures/setup-config fixtures/setup-db)
+(use-fixtures :each fixtures/clear-db)
+
+(deftest create-form
```
Comment (govindkrjoshi) [inline test/bugle_forms/handlers/form_test.clj:15]: Add tests for negative cases

File: test/bugle_forms/handlers/form_test.clj [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                   :session {:user {:name (:user/name user)
+                                    :uuid id}}}
+          response (sut/create request)]
+      (is (= 303 (:status response)))
+      (is (.startsWith (get (:headers response) "Location") "/form-builder/"))
+      (is (not-empty (sql/find-by-keys db/datasource :form
+                                       {:name form-name
+                                        :owner id}))))))
+
+(deftest form-builder
```
Comment (govindkrjoshi) [inline test/bugle_forms/handlers/form_test.clj:30]: Same here

File: test/bugle_forms/model/form_test.clj [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+   [uuid gen/uuid]
+   (let [b64-str (sut/uuid->base64 uuid)
+         uuid-from-b64 (sut/base64->uuid b64-str)]
+     (and
+      (uuid? uuid-from-b64)
+      (= uuid-from-b64 uuid)
+      ;; If we can decode it as base64, it is valid base64
+      (.decode (Base64/getUrlDecoder) b64-str)))))
+
+(deftest create-form
```
Comment (govindkrjoshi) [inline test/bugle_forms/model/form_test.clj:17]: Add tests for negative cases

File: test/bugle_forms/factories.clj [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+   [bugle-forms.specs :as specs]
+   [clojure.spec.alpha :as s]
+   [clojure.spec.gen.alpha :as gen]))
+
+(defn user
+  "Randomly generate a user.
+  Optionally takes a map of overrides."
+  ([] (user {}))
+  ([overrides]
+   (gen/generate
```
Comment (govindkrjoshi) [inline test/bugle_forms/factories.clj:12]: [Note] It's better to return a generator from the factory instead of the realized value using `gen/generate`. This will allow you to compose these factories easily and make bigger factories from it.

**Review by govindkrjoshi — COMMENTED** [review]
Left some comments inline.

**Review by govindkrjoshi — APPROVED** [review]

File: test/bugle_forms/model/form_test.clj [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+  (testing "`get-forms` for a user that does not exist returns empty collection"
+    (let [uuid (gen/generate gen/uuid)]
+      (is (empty? (sut/get-forms {:uuid uuid})))
+      (is (empty? (sut/get-forms {:uuid nil}))))))
+
+;; This is in a separate `deftest` because our fixtures lead to these tests
+;; happening in a rollback-only transaction. Since this test throws an
+;; exception, it causes subsequent tests to fail because the transaction fails,
+;; and thus none of the other database commands are issued.
```
Comment (prabhanshuguptagit) [inline test/bugle_forms/model/form_test.clj:62]: Wouldn't the transaction be rolled back when there's a database exception anyways?

Comment (tfidfwastaken) [reply test/bugle_forms/model/form_test.clj:62]: Yes?

Basically if I stuck this test in between other tests in the `deftest` above, the subsequent tests won't work because the database commands are not issued anymore.
